module5/lessons/Week1/RefactoringPatterns.md

```diff
@@ -0,0 +1,415 @@
+    void Pause();
+    void Stop();
+    void AdjustVolume();
+    void ShowSubtitle();
+}
+```
+
+We _can_ use this interface to drive the implementation of a Video Player (like youtube), and Audio Player (like spotify), and a Slide Player (like a carousel of images on a web-page).  But, we will need to be very careful with how we implement each of these classes:
+
+```c#
+public class VideoPlayer : IMediaPlayer
+{
+    public void Play()
+    {
+        // Play video
+    }
+
+    public void Pause()
+    {
+        // Pause video
+    }
+
+    public void Stop()
+    {
+        // Stop video
+    }
+
+    public void AdjustVolume()
+    {
+        // Adjust volume for video
+    }
+
+    public void ShowSubtitle()
+    {
+        // Show subtitle for video
+    }
+}
+```
+
+So far, so good...
+
+```c#
+public class AudioPlayer : IMediaPlayer
+{
+    public void Play()
+    {
+        // Play audio
+    }
+
+    public void Pause()
+    {
+        // Pause audio
+    }
+
+    public void Stop()
+    {
+        // Stop audio
+    }
+
+    public void AdjustVolume()
+    {
+        // Adjust volume for audio
+    }
+
+    public void ShowSubtitle()
+    {
+        // Audio player doesn't need subtitle functionality
+        throw new NotImplementedException("AudioPlayer doesn't support subtitles.");
+    }
+}
+```
+
+Our AudioPlayer is now getting a bit more complicated...
+
+```c#
+public class SlidePlayer : IMediaPlayer
+{
+    public void Play()
+    {
+        // Play streaming content
+    }
+
+    public void Pause()
+    {
+        // Pause streaming content
+    }
+
+    public void Stop()
+    {
+        // Stop streaming content
+    }
+
+    public void AdjustVolume()
+    {
+        // Slide player doesn't need adjust volume functionality
+        throw new NotImplementedException("SlidePlayer doesn't support volume.");
+    }
+
+    public void ShowSubtitle()
+    {
+        // Streaming player doesn't need subtitle functionality
+        throw new NotImplementedException("StreamingPlayer doesn't support subtitles.");
+    }
+}
+```
+
+And now our Slide Player is just a mess.
+
+There are really only 3 common methods for all of these players, so it would make the most sense to break up our IMediaPlayer into 3 seperate interfaces:
+
+```c#
+public interface IPlayable
+{
+    void Play();
+    void Pause();
+    void Stop();
+}
+
+public interface IVolumeAdjustable
+{
+    void AdjustVolume();
+}
+
+public interface ISubtitleSupport
+{
+    void ShowSubtitle();
+}
+
+public class VideoPlayer : IPlayable, IVolumeAdjustable, ISubtitleSupport
+{
+    // Implement relevant methods
+}
+
+public class AudioPlayer : IPlayable, IVolumeAdjustable
+{
+    // Implement relevant methods
+}
+
+public class SlidePlayer : IPlayable
+{
+    // Implement relevant methods
+}
+```
+
+With the **Extract Interfaces** method, we have smaller, more re-usable interfaces!  This also adheres better to the Interface Segregation principle.
+
+## Move Method
```

Comment:
What do you think about at the end of each Refactoring Pattern asking which SOLID Principle this refactor feels like it's most helping with?

For me, it feels like 

Extract Method, single responsibility
Extract Interfaces, interface segregation
Move Method, a bit of a stretch but sort of Dependency Inversion Principle